WIP: customize tokenization for path fields

mistydemeo opened the pull request:
The default tokenization used by Elasticsearch doesn't deal well with the ways that path strings are constructed; substring searches often fail to bring up files and SIPs which would be expected by users. For example, trying to search for "events" will fail to bring up files in a SIP named "Events_2005-d1c6c1a2-f613-497f-a527-64eb6861f03c".

This adds several new custom analyzers to perform tokenization, which split strings on path delimeters and on logical word separators within strings (such as `-` and `_`). Fields which need these analyzers are now created as multifields, allowing searches for those fields to attempt to match all of the defined subfields. This WIP PR tests this by defining `transferfile`'s `relative_path` field as a multifield with unanalyzed, path-analyzed, and word-analyzed copies of the field.

mistydemeo committed:
elasticSearchFunctions: customize tokenization for paths

This updates tokenization on the relative_path field in order to
ensure that substring searches work more in line with what users
expect.

This sets up relative_path as a multifield with three fields:

* a raw, unanalyzed copy
* a copy split on path elements (such as /)
* a copy split on word elements (such as _, -)

mistydemeo committed:
elasticSearchFunctions: only run connect_and_create_index once

Hwesta commented on src/archivematicaCommon/lib/elasticSearchFunctions.py:
```diff
@@ -61,6 +61,19 @@
+    'word': {
+        'type': 'pattern',
+        # splits on path delimiters, underscores, and dashes
+        'pattern': '(\\/|\\\|_|-)'
```
Also split on `.` maybe?

Hwesta commented on src/archivematicaCommon/lib/elasticSearchFunctions.py:
```diff
@@ -206,20 +219,13 @@ def aip_mapping_is_correct(conn):
-    else:
-        conn = False
-
+def connect_and_create_index(index):
```
Why remove the retry?

mistydemeo replied:
As written, the original version could never have worked as expected. It appeared to rely on `put_mapping` silently failing, which may or may not have been possible with pyes but which is definitely not possible in the current configuration.

Hwesta replied:
Ahh, that makes sense.

Hwesta commented on src/archivematicaCommon/lib/elasticSearchFunctions.py:
```diff
@@ -206,20 +219,13 @@ def aip_mapping_is_correct(conn):
+    conn.indices.create(index, ignore=400)
+    # Always put the mapping, even if the index already exists, because
+    # there may have been changes to the mapping structure; rerunning this
+    # will allow additional fields to be added to the index
```
Update put_mapping to `ignore=[400,404]` as well?

mistydemeo replied:
What's the circumstances in which `404` might happen when creating the index?

Hwesta replied:
Heh, not sure.  I was looking at http://elasticsearch-py.readthedocs.org/en/master/api.html?highlight=ignore#ignore which suggested 404 might be useful, but possibly not in this context.

Hwesta commented:
I like this idea, and I'd like to see the better analyzers on other fields.

mistydemeo committed:
elasticSearchFunctions: customize AIP path tokenization, too

mistydemeo committed:
Search relative_path subfields by name

Searching these fields by name allows match queries to correctly compare
against all of the analyzed fields.

mistydemeo committed:
Backlog search: add path as a search type

mistydemeo committed:
Backlog search: specify index/doc_type

These are needed in order to look up subfields.

mistydemeo commented:
Added several new commits to update how queries are constructed.

After quite a bit of testing, I determined that `match` queries weren't behaving well on these multifields; they were generally matching against only one of the subfields, which meant that queries that should have been valid weren't producing expected results. For example, matching a term against `relative_path` or `relative_path.*` returned no results, whereas matching it against `relative_path.word` returned the desired result.

We already have some support in our query generation code to expand queries on object fields into [multi_match](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-multi-match-query.html) queries on the string subfields of those objects. I've updated this to support path fields like `relative_path`, and added a constant to `elasticSearchFunctions` which explicitly marks the subfields so we don't have to query the index for the structure in this case. This means queries on path fields go from looking like this:

``` json
{
  "filter": {
    "term": {
      "status": "backlog"
    }
  },
  "query": {
    "bool": {
      "must": [],
      "must_not": [],
      "should": [
        {
          "multi_match": {
            "query": "multimedia",
            "fields": [
              "relative_path.*"
            ]
          }
        }
      ]
    }
  }
}
```

To looking like this:

``` json
{
  "filter": {
    "term": {
      "status": "backlog"
    }
  },
  "query": {
    "bool": {
      "must": [],
      "must_not": [],
      "should": [
        {
          "multi_match": {
            "query": "multimedia",
            "fields": [
              "relative_path.raw",
              "relative_path.path",
              "relative_path.word"
            ]
          }
        }
      ]
    }
  }
}
```

With the latter query returning results which the former query will not. I haven't yet encountered any cases where this returns undesirable results, or otherwise has negative impact on existing queries.

I've also updated our search form JS to permit "path" as a type, which queries `relative_path` directly. We previously only had filename searches, which isn't sufficient for what we're doing here. Based on discussions with @jhsimpson, I think we should promote that to the default search field instead of "any".

mistydemeo commented:
Deployed to missy for QA.

mistydemeo commented:
Rebased on the new ES changes.

sevein commented:
I'm closing this PR as Misty is not actively working on this anymore - the branch will be maintained. I've filed a new issue to track the problem that she described - https://github.com/archivematica/Issues/issues/144. Thanks!
